apps/system/js/app_text_selection_dialog.js

```diff
@@ -0,0 +1,472 @@
+
+  // Distance between selected area and the bottom of menu when menu show on
+  // the top of selected area.
+  // By UI spec, 12px from the top of dialog to utility menu.
+  AppTextSelectionDialog.prototype.DISTANCE_FROM_MENUBOTTOM_TO_SELECTEDAREA =
+    12;
+
+  // Distance between selected area and the top of menu when menu show on
+  // the bottom of selected area.
+  // caret tile height is controlled by gecko, we estimate the height as
+  // 22px. So 22px plus 12px which defined in UI spec, we get 34px from
+  // the bottom of selected area to utility menu.
+  AppTextSelectionDialog.prototype.DISTANCE_FROM_SELECTEDAREA_TO_MENUTOP = 43;
+
+  // Minimum distance between bubble and boundary.
+  AppTextSelectionDialog.prototype.DISTANCE_FROM_BOUNDARY = 5;
+
+  AppTextSelectionDialog.prototype.ID_NAME = 'AppTextSelectionDialog';
+
+  AppTextSelectionDialog.prototype.ELEMENT_PREFIX = 'textselection-dialog-';
+
+  AppTextSelectionDialog.prototype.start = function tsd_start() {
+    if (this._enabled) {
+      return;
+    }
+    this._enabled = true;
+    if (this.app) {
+      this.app.element.addEventListener('mozbrowsercaretstatechanged', this);
+    } else {
+      window.addEventListener('mozChromeEvent', this);
+    }
+  };
+
+  AppTextSelectionDialog.prototype.stop = function tsd_stop() {
+    if (!this._enabled) {
+      return;
+    }
+    this._enabled = false;
+    if (this.app) {
+      this.app.element.removeEventListener('mozbrowsercaretstatechanged', this);
+    } else {
+      window.removeEventListener('mozChromeEvent', this);
+    }
+  };
+
+  AppTextSelectionDialog.prototype.debug = function tsd_debug(msg) {
+    if (DEBUG || this._DEBUG) {
+      console.log('[Dump: ' + this.ID_NAME + ']' +
+        JSON.stringify(msg));
+    }
+  };
+
+  AppTextSelectionDialog.prototype.handleEvent = function tsd_handleEvent(evt) {
+    switch (evt.type) {
+      case 'mozbrowsercaretstatechanged':
+        evt.preventDefault();
+        evt.stopPropagation();
+        this._handleCaretStateChanged(evt.detail);
+        break;
+      case 'mozChromeEvent':
+        switch (evt.detail.type) {
+          case 'caretstatechanged':
+            evt.preventDefault();
+            evt.stopPropagation();
+            this._handleCaretStateChanged(evt.detail.detail);
+            break;
+        }
+    }
+  };
+
+  AppTextSelectionDialog.prototype._handleCaretStateChanged =
+    function tsd__handleCaretStateChanged(detail) {
+      if (!detail) {
+        return;
+      }
+      this.debug('on receive caret state change event');
+      this.debug(JSON.stringify(detail));
+      if (!this._injected) {
+        this.render();
+        this._injected = true;
+      }
+      if (detail.reason === 'visibilitychange' && !detail.caretVisibe) {
+         this.hide();
+         return;
+      }
+      if (detail.reason === 'presscaret') {
+        this.hide();
+        return;
+      }
+      if (!detail.selectionVisible && !detail.caretVisible) {
```

Comment:
Please use if (!detail.selectionVisible) here. Otherwise, the text selection dialog will still be visible when the whole selection is being scrolled out of screen.